### ISSUE TITLE
Devs using C++17 features will break internal builds

### ISSUE THREAD

**Issue by ZainRizvi** [pytorch/pytorch#96427]
**The ask:**
Make pytorch's fbcode builds and github builds have the same set of C++ features available. Either by preventing their usage on github, or enabling them internally. Otherwise non-Meta devs have to go through multiple rounds of their PR getting reverted before it can finally be landed

From @amjames 

https://github.com/pytorch/pytorch/pull/94048 was reverted because the use of C++17 features broke an internal work stream. It is not blocked indefinitely the internal team is using an older version of MSVC which does not fully support all C++ 17 features. Without having that specific MSVC version in pytorch CI we can’t debug the issue. All CI on our side is green so we have no signal to use to test the compatibility needed here.

This has been brought up at a few levels, we have spoken to @malfet, I have mentioned it to @albanD and Joe as well, so a lot of people are aware of this, but all of them have stated we will need to route this internally to resolve it. The main issue here is this: PyTorch now ostensibly supports C++17, however internal workstreams using out of date compilers can lead to PRs getting reverted like this. Without any signal to use to fix the issue we have no path to resolution.
The internal team either needs to migrate to a compatible compiler or we need to roll back support on the PyTorch side. As long as this situation continues it will make a lot of the GH1 work moot.

The underlying issue is also preventing us from landing a slew of performance improvements related to sparse. We don’t want to try to send them through and get them reverted as we have had complaints due to a lot of reverts on this contributor’s PRs lately. This stack https://github.com/pytorch/pytorch/pull/94401 uses similar techniques and we do not want to try and land anything until we get resolution on the first issue, it is suspected that we will have them reverted for the same reason, but we can’t verify that a priori without a signal.

cc @malfet @seemethere

**Comment by amjames** [comment]
cc @nikitaved @cpuhrsch

@nikitaved good place to put details of specific things we are stumbling over over

**Comment by cyyever** [comment]
If I don't understand it wrongly. The whole point is that the buggy C++ infrastructure inside META slows down the adoption of C++17 and possible performance gains in pytorch code base and no META dev here has the necessary resources to upgrade  or make a clean plan to upgrade the in-house MSVC compiler in the foreseeable future.

From the respect of software engineering, the best option is to cut down the connection between the open sourced version and META version of pytorch. Our contributors don't need to care about META things and the open sourced version can evolve more quickly and independently. Up to some point the divergence is large enough that META may want to import the changes, and that would push them to upgrade their infrastructure.

However, in reality, pytorch was developed and open sourced by META. The whole ecosystem benefits from this decision. I appreciate their contributions and accept other options discussed here.

**Comment by nikitaved** [comment]
So, yeah, luckily I just downgraded my code to cpp14. It is not the first time I hit issues with MSVC, it has problems understanding nested code inside macro expansions. And nesting macros and lambdas is most likely a recipe for disaster. That holds for cpp17 features inside macros as well.

**Comment by malfet** [comment]
Internal use of VS2017 is being deprecated internally, but otherwise we should be modern compiler. There used to be a glitch when some parts of the code were compiled with `-std=c++14`, but this also should all be eliminated. 
@nikitaved how recent was this revert? As last reason for revert on https://github.com/pytorch/pytorch/pull/94048 seems to be `signed-unsigned compare` which is not specific to C++ standard (but also should have been caught by OSS CI)

**Comment by soumith** [comment]
> From the respect of software engineering, the best option is to cut down the connection between the open sourced version and META version of pytorch. Our contributors don't need to care about META things and the open sourced version can evolve more quickly and independently.

I thought it is worth pointing out that we don't live in an isolated world. PyTorch is a pragmatic ecosystem -- every stakeholder that brings a big chunk of value to the table is considered with care. Some of these are pure userland considerations, such as whether it is Ubuntu 18.04 users beyond EOL cutoff (we supported various distros until we saw users stopped downloading from those distros), Python 2 users while Python 3 was the main version (we supported Python 2 until we saw downloads for Python 2 start dwindling). There is a correlation between "We care about users" and "users like PyTorch".
Some of these users cross the boundary into developers (and funding). We take several constraints, from Meta, Microsoft, NVIDIA, AMD, Google and others in light of the fact that they fund development -- pay salaries, give money for CI infrastructure, etc. Meta is pretty evidentially the largest funder of PyTorch. The consequences of cutting the connection between the OSS version and the Meta version -- without any hint of nefariousness or bad intent -- can result in reduced development of the OSS side purely from the practical consequences of such a structural change. Basically, engineering time will then have to be split (over time) between two versions, and if a small set of decisions in a long time horizon go in an incompatible direction then it can result in permanent downfunding of the OSS side. We take the constraints and considerations coming from several corporate funders pretty seriously, because it has a meaningful effect on the development of PyTorch in open-source. In my opinion, this all is an acceptable price to pay.
